src/matrix/distribution.cpp

```diff
@@ -98,85 +100,61 @@ Distribution::Distribution(Distribution rhs, const SubDistributionSpec& spec)
 
-  // The local size is computed in the following way:
-  // If the last element belongs to my rank:
-  //   local_size = (local_nr_tiles - 1) * tile_size + size of last tile.
-  // otherwise:
-  //   local_size = local_nr_tiles * tile_size
-  // Additionally, if the first element belongs to my rank and there is at least
-  // one tile subtract the offset because it was added in the first step to
-  // temporarily assume the first block is complete:
-  //   local_size = local_size - offset
   SizeType row = 0;
   if (size_.rows() > 0) {
-    if (rank_index_.row() == rankGlobalTile(global_nr_tiles_.rows() - 1, tilesPerBlock<Coord::Row>(),
-                                            grid_size_.rows(), source_rank_index_.row())) {
-      const auto last_tile_rows = (size_.rows() + offset_.row() - 1) % tile_size_.rows() + 1;
-      row = (tile_row - 1) * tile_size_.rows() + last_tile_rows;
-    }
-    else {
-      row = tile_row * tile_size_.rows();
-    }
-    if (tile_row > 0 && rank_index_.row() == source_rank_index_.row()) {
-      DLAF_ASSERT(row >= globalTileElementOffset<Coord::Row>(), row,
-                  globalTileElementOffset<Coord::Row>());
-      row -= globalTileElementOffset<Coord::Row>();
-    }
+    // Start from full tiles
+    row = tile_row * tile_size_.rows();
+
+    // Fix first tile size removing local offset
+    row -= local_tile_element_offset<Coord::Row>();
+
+    // Fix last tile size
+    if (rank_index_.row() == rank_global_tile<Coord::Row>(nr_tiles_.rows() - 1))
+      // remove the elements missing in the last tile
+      row -= nr_tiles_.rows() * tile_size_.rows() -
+             (size_.rows() + global_tile_element_offset<Coord::Row>());
   }
   SizeType col = 0;
   if (size_.cols() > 0) {
-    if (rank_index_.col() == rankGlobalTile(global_nr_tiles_.cols() - 1, tilesPerBlock<Coord::Col>(),
-                                            grid_size_.cols(), source_rank_index_.col())) {
-      const auto last_tile_cols = (size_.cols() + offset_.col() - 1) % tile_size_.cols() + 1;
-      col = (tile_col - 1) * tile_size_.cols() + last_tile_cols;
-    }
-    else {
-      col = tile_col * tile_size_.cols();
-    }
-    if (tile_col > 0 && rank_index_.col() == source_rank_index_.col()) {
-      DLAF_ASSERT(col >= globalTileElementOffset<Coord::Col>(), col,
-                  globalTileElementOffset<Coord::Col>());
-      col -= globalTileElementOffset<Coord::Col>();
-    }
+    // Start from full tiles
+    col = tile_col * tile_size_.cols();
+
+    // Fix first tile size removing local offset
+    col -= local_tile_element_offset<Coord::Col>();
+
+    // Fix last tile size
+    if (rank_index_.col() == rank_global_tile<Coord::Col>(nr_tiles_.cols() - 1))
+      // remove the elements missing in the last tile
+      col -= nr_tiles_.cols() * tile_size_.cols() -
+             (size_.cols() + global_tile_element_offset<Coord::Col>());
   }
 
   // Set local_size_.
   local_size_ = LocalElementSize(row, col);
```

Comment:
Would it make sense to factor out a helper which is templated on the `Coord` just to avoid repeating the same code for `Row` and `Col`?